### PR TITLE
Add predefined types for tagged int8 and int16

### PR DESCRIPTION
**Pull request by jvanburen**
I also created modules for them in `Stdlib_beta`

### CODE REVIEW

**Comment by jvanburen** [comment]
This will be a prerequisite for unboxed int8 and int16 types

File: testsuite/tests/lib-smallint/test_int8.ml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+let test_logical1 = test_conv1 ~equal:equal_logical
+
+let test_logical2 = test_conv2 ~equal:equal_logical
+
+let reference_shift_right_logical x i =
+  (* we need to ensure that we shift in zero bytes, which is incompatible with
+     sign-extension *)
+  Int.shift_right_logical (if i > 0 then x land mask else x) i
+
+let () =
```
Comment (ccasin) [inline testsuite/tests/lib-smallint/test_int8.ml:90]: Things defined in the library that aren't tested here: `neg`, `min_int`, and `max_int`.

Testing `neg` seems straightforward.  For `min_int` and `max_int` I guess you can just compare them to the concrete expected values.

File: testsuite/tests/lib-smallint/test_int8.ml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  test_arith1 Int8.pred Int.pred;
+  test_arith1 Int8.abs Int.abs;
+  test_logical2 Int8.logand Int.logand;
+  test_logical2 Int8.logor Int.logor;
+  test_logical2 Int8.logxor Int.logxor;
+  test_logical1 Int8.lognot Int.lognot;
+  for shift = 0 to int_size do
+    let apply_shift f x = f x shift in
+    test_logical1 (apply_shift Int8.shift_right) (apply_shift Int.shift_right);
+    test_conv1
```
Comment (ccasin) [inline testsuite/tests/lib-smallint/test_int8.ml:110]: Can this be `test_logical`?

File: testsuite/tests/lib-smallint/test_int8.ml [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  test_arith1 Int8.succ Int.succ;
+  test_arith1 Int8.pred Int.pred;
+  test_arith1 Int8.abs Int.abs;
+  test_logical2 Int8.logand Int.logand;
+  test_logical2 Int8.logor Int.logor;
+  test_logical2 Int8.logxor Int.logxor;
+  test_logical1 Int8.lognot Int.lognot;
+  for shift = 0 to int_size do
+    let apply_shift f x = f x shift in
+    test_logical1 (apply_shift Int8.shift_right) (apply_shift Int.shift_right);
```
Comment (ccasin) [inline testsuite/tests/lib-smallint/test_int8.ml:109]: Why is `test_logical` used for the arithmetic shift?

**Review by ccasin — COMMENTED** [review]
Looks good, mostly minor suggestions.

File: otherlibs/stdlib_beta/int16.ml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+let[@inline always] rem x y = of_int (Int.rem (to_int x) (to_int y))
+
+let[@inline always] succ x = of_int (Int.succ (to_int x))
+
+let[@inline always] pred x = of_int (Int.pred (to_int x))
+
+let[@inline always] abs x = of_int (Int.abs (to_int x))
+
+let[@inline always] logand x y =
+  unsafe_of_int (Int.logand (to_int x) (to_int y))
```
[the remark this answers is not in the document]
Comment (jvanburen) [reply otherlibs/stdlib_beta/int16.ml:62]: Also I should check that abs works at min_int

File: otherlibs/stdlib_beta/int16.mli [hunk 5]
```diff
@@ -0,0 +1,136 @@
+(**************************************************************************)
```
[the remark this answers is not in the document]
Comment (jvanburen) [reply otherlibs/stdlib_beta/int16.mli:1]: of_string is explicitly commented out of int.mli, so I didn't add that , but I added the hash functions and some others

File: testsuite/tests/lib-smallint/test_int8.ml (hunk 3 above)
Comment (jvanburen) [reply testsuite/tests/lib-smallint/test_int8.ml]: it's for the right shifts, not just the arithmetic one

File: otherlibs/stdlib_beta/int.ml [hunk 6]
```diff
@@ -0,0 +1,319 @@
+(**************************************************************************)
```
Comment (ccasin) [inline otherlibs/stdlib_beta/int.ml:1]: I think this module should be named something other than Int.  It's too confusing and error-prone to have two different modules named `Int` in the stdlib, even if one is in `stdlib_beta`.  One day we'd like all of this to be in the normal upstream stdlib!

File: otherlibs/stdlib_beta/.ocamlformat-enable [hunk 7]
```diff
@@ -1,3 +1,7 @@
 stdlib_beta.mli
 stdlib_beta.ml
-
+int_intf.ml
```
Comment (ccasin) [inline otherlibs/stdlib_beta/.ocamlformat-enable:3]: I don't see a file called `int_intf.ml` anywhere.

File: otherlibs/stdlib_beta/int.ml [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+      (Container.shift_left (inject t) unused_bits)
+      unused_bits
+
+  let zero = sign_extend Container.zero
+
+  let one = sign_extend Container.one
+
+  let minus_one = sign_extend Container.minus_one
+
+  let[@inline available] neg x = sign_extend (Container.neg (inject x))
```
Comment (ccasin) [inline otherlibs/stdlib_beta/int.ml:228]: I think these should just be marked "inline" - we proactively want them to be inlined, not just if the heuristic chooses it.  But probably you had some reason in mind I'm missing?

**Review by ccasin — APPROVED** [review]
A few more comments, but eagerly approving now.  Thanks!

File: otherlibs/stdlib_beta/int.ml (hunk 8 above)
Comment (jvanburen) [reply otherlibs/stdlib_beta/int.ml:228]: I was copying what int32.ml had, still inline always?

Comment (jvanburen) [reply otherlibs/stdlib_beta/int.ml]: we'll ask mark and can change later, merging as-is

File: otherlibs/stdlib_beta/int.ml (hunk 6 above)
Comment (jvanburen) [reply otherlibs/stdlib_beta/int.ml]: Changed to `Stdlib_beta.Int_wrapper`

**Pull request merged**